### DISCUSSION TITLE
[RFC] 021 - 自定义助手开场引导

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#891] [RFCs]
## 背景

随着助手越来越多，和用户的使用深入，我们需要给用户提供更加丰富的助手自定义能力。其中开场引导则是一个有意义的部分。它可以帮助用户了解助手的能力，或者快速开始一个会话。

- #890

同时，现在的引导话题还存在操作体验不佳的情况
- #395

## 解决方案

参考 #890 的建议，支持用户自定义开场白，同时可以自定义开场快捷提问。并且可以兼容到未来 #869 的使用。

## 技术实现

新增一种 role=guide 的 message type，用于专门承载开场白类型的消息。

开场白类型的消息支持如下数据结构： 
1. openMessage: 欢迎描述，如果没有填写，则使用系统内置的
2. openQuestions: 开场提问，放一些常见的问题，帮助用户快速开始会话
3. openTopics: 开场的话题模板，使用模板快速开始一组会话。对于场景化的消息来说会非常提效。

## 进展

- 预设 Inbox 的 Welcome Guide #2086

**Comment by kiner-tang** [comment]
这个任务我来跟一下吧

**Reply by arvinxx** [reply]
可以在这个 RFC 下把详细思路先写一下，看看哪里有不清楚的地方

**Reply by kiner-tang** [reply]
好的，我这几天先整理一下

**Comment by kiner-tang** [comment]
想要明确一下需求：
我们实际上是想要在这里增加一个 `guide` 的类型，
![image](https://github.com/lobehub/lobe-chat/assets/10286961/1e792ef3-31c9-4eb4-ac47-4b6ee6769bda)
![image](https://github.com/lobehub/lobe-chat/assets/10286961/3967f52b-3fa0-419a-adf0-0c8e82672044)
并在这边使用 `guide` 类型的消息逻辑替换原本的逻辑是么：
<img width="955" alt="image" src="https://github.com/lobehub/lobe-chat/assets/10286961/64fc2007-8e38-48e7-bccd-3d949ff211ba">

**Reply by arvinxx** [reply]
这是第一步。然后后面针对 guide 类型的消息，就可以实现更多程序化的扩展，例如上面提到的让用户自定义 agent 的欢迎引导

**Reply by kiner-tang** [reply]
按照我的理解，是不是这三个实际上是交由用户自定义配置的：
openMessage: 欢迎描述，如果没有填写，则使用系统内置的
openQuestions: 开场提问，放一些常见的问题，帮助用户快速开始会话
openTopics: 开场的话题模板，使用模板快速开始一组会话。对于场景化的消息来说会非常提效。

**Reply by arvinxx** [reply]
> 按照我的理解，是不是这三个实际上是交由用户自定义配置的： openMessage: 欢迎描述，如果没有填写，则使用系统内置的 openQuestions: 开场提问，放一些常见的问题，帮助用户快速开始会话 openTopics: 开场的话题模板，使用模板快速开始一组会话。对于场景化的消息来说会非常提效。

对，这个需要在 agent 的 settings 里新增相应的配置项。我建议从项目熟悉的角度来说，可以分两个 PR 走：
1. 新增 role=guide 的 message，然后改造相应实现，这样就能直接解决 #395 这个问题；
2. 在助手设置那里，新增 openMessage、openQuestions 两个字段，支持用户配置，实现 #890 的需求

**Reply by kiner-tang** [reply]
好的

**Comment by kiner-tang** [comment]
在 ui 库那边得先支持一下配置是否点击链接在新窗口打开：https://github.com/lobehub/lobe-ui/pull/112

**Reply by arvinxx** [reply]
应该不需要这么做，可以直接针对role=guide 新增一种渲染方式。然后直接在这个renderItem里自定义相关逻辑

**Reply by kiner-tang** [reply]
好的，我看了一下，应该就是在这里增加一个 guide 的一个自定义渲染器吧
<img width="1686" alt="image" src="https://github.com/lobehub/lobe-chat/assets/10286961/15308eab-d98c-47a6-9713-33f77822bdfa">

**Reply by arvinxx** [reply]
对
